toolchain/check/handle_impl.cpp

```diff
@@ -34,34 +38,143 @@ auto HandleImplForall(Context& context, Parse::ImplForallId parse_node)
 
 auto HandleTypeImplAs(Context& context, Parse::TypeImplAsId parse_node)
     -> bool {
-  auto self_id = context.node_stack().PopExpr();
-  context.node_stack().Push(parse_node, self_id);
+  auto [self_node, self_id] = context.node_stack().PopExprWithParseNode();
+  auto self_type_id = ExprAsType(context, self_node, self_id);
+  context.node_stack().Push(parse_node, self_type_id);
+  // TODO: `Self` should come into scope here, at least if it's not already in
+  // scope. Check the design for the latter case.
   return true;
 }
 
-auto HandleDefaultSelfImplAs(Context& /*context*/,
-                             Parse::DefaultSelfImplAsId /*parse_node*/)
-    -> bool {
-  // TODO: Determine self_id and push it onto node stack.
+// If the specified name scope corresponds to a class, returns the corresponding
+// class declaration.
+// TODO: Should this be somewhere more central?
+static auto AsClassScope(Context& context, SemIR::NameScopeId scope_id)
+    -> std::optional<SemIR::ClassDecl> {
+  if (!scope_id.is_valid()) {
+    return std::nullopt;
+  }
+  auto& scope = context.name_scopes().Get(scope_id);
+  if (!scope.inst_id.is_valid()) {
+    return std::nullopt;
+  }
+  return context.insts().TryGetAs<SemIR::ClassDecl>(scope.inst_id);
+}
+
+auto HandleDefaultSelfImplAs(Context& context,
+                             Parse::DefaultSelfImplAsId parse_node) -> bool {
+  // Find the enclosing scope for the `impl`.
+  // TODO: Do this without modifying the node stack.
```

Comment:
Is this something you're actively working on? Rather than pushing the enclosing_scope_id onto the stack (which seems to be resulting in NameScopeId support being added to NodeStack IIUC) would a `PeekScope(1)` work? Could there be an intervening scope which would interfere with that? The `current_scope` API is maybe a little restrictive; perhaps that should be more accessible in general.

Reply:
This is cleaned up in #3691.